src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionKey.cs

```diff
@@ -6,35 +6,41 @@ namespace System.Net.Http
 {
     internal readonly struct HttpConnectionKey : IEquatable<HttpConnectionKey>
     {
-        public readonly bool UsingSSL;
         public readonly string Host;
         public readonly int Port;
+        public readonly string SslHostName;     // null if not SSL
 
-        public HttpConnectionKey(Uri uri)
+        public HttpConnectionKey(string host, int port, string sslHostName)
         {
-            UsingSSL =
-                HttpUtilities.IsSupportedNonSecureScheme(uri.Scheme) ? false :
-                HttpUtilities.IsSupportedSecureScheme(uri.Scheme) ? true :
-                throw new ArgumentException(SR.net_http_client_http_baseaddress_required, nameof(uri));
-
-            Host = uri.Host;
-            Port = uri.Port;
+            Host = host;
+            Port = port;
+            SslHostName = sslHostName;
         }
 
+        // The common cases here are SslHostName == null or SslHostName == Host
         public override int GetHashCode() =>
-            UsingSSL.GetHashCode() << 16 ^ Host.GetHashCode() ^ Port.GetHashCode();
+            SslHostName == null ? Host.GetHashCode() ^ Port.GetHashCode() :
+            SslHostName == Host ? Host.GetHashCode() ^ Port.GetHashCode() ^ 0x1 :
+            Host.GetHashCode() ^ Port.GetHashCode() ^ SslHostName.GetHashCode();
 
         public override bool Equals(object obj) =>
             obj != null &&
             obj is HttpConnectionKey &&
             Equals((HttpConnectionKey)obj);
 
         public bool Equals(HttpConnectionKey other) =>
-            UsingSSL == other.UsingSSL &&
             Host == other.Host &&
-            Port == other.Port;
+            Port == other.Port &&
```

Comment:
I am assuming that this is the socket port, if it is I will tend to check it before the other ones.

Reply:
Why?  Because the equality test is cheaper?  Unfortunately, Port is very likely to be either 80 or 443 so the equality test is not super useful.

Reply:
Ops... I was thinking in the likely hood of them being different but here the port is not (duh!) the socket port...